Feat/contract page transactions

Benjtalkshow opened the pull request:
close #41
## Contract page transactions
This PR implements a new feature to fetch and display transactions specific to a given contract address.
 
###  New Features
- Introduced `fetchTransactionsData` function to retrieve transactions from the database
- Created `AddressTransactions` component to display transaction data
- Implemented filtering for valid transactions (those with receipts)
- Added error handling for transaction processing to improve reliability
- Ensured accurate counting of valid transactions

### Technical Details

- Utilizes Prisma for database interactions
- Implements Promise.all for efficient asynchronous processing of transactions
- Limits displayed transactions to 25 for performance, while showing the total count of valid transactions
- Uses the existing `TxsTable` component for consistent transaction display across the application

###  Testing

- Tested with various Ethereum addresses to ensure correct fetching and display of transactions
- Verified error handling by simulating failed transaction processing
- Confirmed accurate counting of valid transactions

### Component Refactoring
- Moved and renamed the latest-transaction-table and its row component to components/lib for better organization and clarity.

### Performance Considerations
- The function fetches all associated transactions before processing, which may impact performance for addresses with a large number of transactions
- Future optimization could include pagination.

Here is the preview below (Desktop)
![image](https://github.com/user-attachments/assets/ccca98d2-89e8-4a02-ac80-2fe2242c9663)
Here is the preview below (Desktop)
![image](https://github.com/user-attachments/assets/b37411d3-10bc-48f2-9afb-f7dcc85d58fb)

I await any corrections. Thanks

Benjtalkshow committed:
Implement contract-specific page transaction fetching and display

Benjtalkshow committed:
Merge branch 'main' of https://github.com/Benjtalkshow/op-scan into feat/contract-page-transactions

Benjtalkshow committed:
Update feat/contract-page-transactions

saimeunt commented on src/components/pages/address/address-transactions.tsx:
```diff
@@ -1,13 +1,66 @@
+import TxsTable from "@/components/lib/txs-table";
 
-const AddressTransactions = ({ address }: { address: Address }) => {
+const TRANSACTION_LIMIT = 25;
```
Use `process.env.NEXT_PUBLIC_TXS_PER_PAGE` instead.

saimeunt commented on src/components/pages/address/address-transactions.tsx:
```diff
@@ -1,13 +1,66 @@
+  const associatedTransactions = await prisma.transaction.findMany({
+    where: {
+      OR: [
+        { from: { contains: address } },
```
I'm curious why is this needed, can't we use `from: address` directly?
If the addresses do not match exactly maybe this is because we store checksum address in the database and the address in the URL should be formatted using viem `getAddress`.

saimeunt commented on src/components/pages/address/address-transactions.tsx:
```diff
@@ -1,13 +1,66 @@
+        { from: { contains: address } },
+        { to: { contains: address } },
+      ],
+    },
```
You're missing a `take: Number(process.env.NEXT_PUBLIC_TXS_PER_PAGE)` here to only fetch the correct amount of transactions.

saimeunt commented on src/components/pages/address/address-transactions.tsx:
```diff
@@ -1,13 +1,66 @@
+  const filteredValidTransactions = validTransactions.filter(tx => tx !== null) as TransactionWithReceipt[];
+  return {
+    transactions: filteredValidTransactions.slice(0, TRANSACTION_LIMIT),
+    totalCount: filteredValidTransactions.length,
```
Total count should be computed in a separate prisma query, see https://www.prisma.io/docs/orm/reference/prisma-client-reference#count

saimeunt commented on src/components/pages/address/address-transactions.tsx:
```diff
@@ -1,13 +1,66 @@
-      <CardHeader>Latest 25 from a total of N transactions</CardHeader>
-      <CardContent>TRANSACTIONS</CardContent>
+      <CardHeader>
+        Latest {transactions.length >= TRANSACTION_LIMIT ? TRANSACTION_LIMIT : transactions.length} from a total of {totalCount} valid transactions
```
You can safely use transactions.length here after making the rest of the requested changes.

saimeunt reviewed (changes requested):
@Benjtalkshow you're almost there, please make the requested changes and we'll merge this!

Benjtalkshow committed:
Optimize AddressTransactions: refactor pagination, address handling, and query performance

Benjtalkshow committed:
Merge branch 'main' of https://github.com/Benjtalkshow/op-scan into feat/contract-page-transactions

Benjtalkshow commented:
@saimeunt 

The following changes have been implemented:

- Replaced `TRANSACTION_LIMIT` constant with `process.env.NEXT_PUBLIC_TXS_PER_PAGE` for consistency.
- Simplified the address query by using the checksum address directly, assuming addresses are stored in checksum format in the database.
- Added a take limit to the transaction query to fetch only the required number of transactions.
- Separated the total count query for better performance, using Prisma's count method.
- Updated the component to use `transactions.length` for displaying the number of fetched transactions.

The component now correctly displays the latest transactions for a given address, with accurate count information.
Please review these changes and let me know if any further adjustments are needed.
